### PR TITLE
[New features] MoE: add AllGather + ReduceScatter dispatcher with intermediate-dim EP sharding

### PR DESCRIPTION
**Pull request by adam-xiaoyao**
### PR Category

Distributed Strategy

### PR Types

New features

### Description

Introduce a new MoE token dispatcher — `allgather` — alongside the existing
`deepep` / `alltoall` / `hybridep` options. Semantic difference:

- Existing dispatchers: each EP rank holds *a subset of experts* at full intermediate width
- AllGather dispatcher: each EP rank holds *every expert*, with the intermediate dim sharded across EP

### 是否引起精度变化

否

### CODE REVIEW

File: src/paddlefleet/transformer/moe/moe_layer.py [hunk 1]
```diff
@@ -638,27 +743,62 @@ def custom_forward(
...
             hidden_states = self.fc1_latent_proj(hidden_states)
+        elif self.use_latent_moe and self._latent_hidden is not None:
+            hidden_states = self._latent_hidden
+            self._latent_hidden = None
 
         should_log_balance = framework._dygraph_tracer()._has_grad
         with profile("dispatch"):
             hidden_states, _ = self.dispatch(
                 hidden_states, probs, routing_map, topk_weights, topk_indices
             )
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/moe_layer.py:825]: 已修改

File: src/paddlefleet/transformer/transformer_config.py [hunk 2]
```diff
@@ -527,6 +533,12 @@ class TransformerConfig(ModelParallelConfig):
     moe_ep_barrier: bool = True
     """Whether to use barrier for expert parallelism."""
 
+    moe_allgather_gate_overlap: bool = False
+    """Whether to overlap AllGather of hidden_states with gate computation in the
+    AllGather MoE dispatcher. When True, the AllGather is issued asynchronously on
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/transformer_config.py]: 不必修改

File: src/paddlefleet/transformer/moe/moe_utils.py [hunk 3]
```diff
@@ -789,3 +789,26 @@ def backward(ctx, grad):
...
+    This is the dual of :class:`AllGatherGroupOp` and is used by the
+    'allgather' MoE token dispatcher to combine partial expert outputs along
+    the EP group: forward sums across EP ranks while scattering along the
+    leading (token) dimension; backward replicates the gradient via
+    all-gather.
+    """
+
+    @staticmethod
+    def forward(ctx, input, group=None):
+        paddle.distributed.barrier(group)
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/moe_utils.py]: 已修改

File: src/paddlefleet/transformer/moe/token_dispatcher.py [hunk 4]
```diff
@@ -1093,3 +1104,505 @@ def combine_postprocess(self, permutated_local_input_tokens: paddle.Tensor):
...
+        ctx.bwf, fn_out = manual_backward(fn, is_first_fwd, *fn_args)
+
+        ctx.group = group
+        ctx.input_shape = list(x.shape)
+        task.wait()
+        return (combined_x,) + fn_out  # noqa: RUF005
+
+    @staticmethod
+    def backward(ctx, grad_output, *fn_out_grads):
+        group = ctx.group
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/token_dispatcher.py:1426]: OK, 我修改。

File: src/paddlefleet/transformer/moe/token_dispatcher.py [hunk 5]
```diff
@@ -835,7 +839,16 @@ def combine_preprocess(self, hidden_states: paddle.Tensor):
...
         )
 
-    def token_combine(self, hidden_states: paddle.Tensor, async_finish=False):
+    def token_combine(
+        self,
+        hidden_states: paddle.Tensor,
+        combine_overlap_handle: dict | None = None,
+        async_finish=False,
+    ):
+        assert combine_overlap_handle is None, (
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/token_dispatcher.py]: OK， 我修改。

File: src/paddlefleet/transformer/moe/token_dispatcher.py [hunk 6]
```diff
@@ -1093,3 +1106,516 @@ def combine_postprocess(self, permutated_local_input_tokens: paddle.Tensor):
...
+        else:
+            global_hidden_states = AllGatherGroupOp.apply(
+                reshaped_input, self.moe_group
+            )
+
+        # Memory-saving fused path: skip permute entirely; the fused SonicMoE
+        # kernels (_UpProjection/_DownProjection) handle gather/scatter
+        # internally. We only AllGather the topk metadata and return the
+        # unpermuted global hidden states.
+        assert topk_indices is not None and topk_weights is not None, (
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/token_dispatcher.py]: OK, 我修改。

File: src/paddlefleet/transformer/moe/token_dispatcher.py [hunk 7]
```diff
@@ -1093,3 +1106,516 @@ def combine_postprocess(self, permutated_local_input_tokens: paddle.Tensor):
...
+    @staticmethod
+    def forward(ctx, x, group, *fn_args, fn, is_first_fwd=False):
+        if group is None or group.nranks == 1:
+            combined_x = x.clone()
+            ctx.bwf, fn_out = manual_backward(fn, is_first_fwd, *fn_args)
+            ctx.group = group
+            return (combined_x,) + fn_out  # noqa: RUF005
+
+        output_shape = list(x.shape)
+        assert output_shape[0] % group.nranks == 0, (
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/token_dispatcher.py]: OK， 我修改。

File: src/paddlefleet/transformer/moe/moe_expert.py [hunk 8]
```diff
@@ -287,7 +300,91 @@ def sharded_state_dict(
...
+
         state_dict = self.state_dict(structured_name_prefix="")
+
+        if is_intermediate_sharded:
+            assert self.ep_group is not None, (
+                "intermediate-sharded layout requires an EP group; "
+                f"got ep_group=None. Check moe_token_dispatcher_type "
+                f"({self.config.moe_token_dispatcher_type!r}) and EP "
+                f"initialisation."
+            )
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/moe_expert.py]: 我修改。

File: src/paddlefleet/transformer/moe/token_dispatcher.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/token_dispatcher.py]: OK， 我修改。

File: src/paddlefleet/transformer/moe/token_dispatcher.py [hunk 9]
```diff
@@ -1093,3 +1107,417 @@ def combine_postprocess(self, permutated_local_input_tokens: paddle.Tensor):
...
+        # previous handle.
+        if self._pre_ag_handle is not None:
+            try:
+                self._pre_ag_handle["task"].wait()
+            except (RuntimeError, OSError) as _e:
+                logger.warning(
+                    "pre_allgather: leftover async task wait failed (%s), "
+                    "discarding handle.",
+                    _e,
+                )
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/token_dispatcher.py:1542]: 这里没有return ou啊

File: src/paddlefleet/transformer/moe/moe_layer.py [hunk 10]
```diff
@@ -613,9 +706,16 @@ def permute(self, hidden_states: paddle.Tensor):
     def unpermute(self, hidden_states: paddle.Tensor):
         return self.token_dispatcher.combine_preprocess(hidden_states)
 
-    def combine(self, hidden_states: paddle.Tensor, async_finish: bool = False):
+    def combine(
+        self,
+        hidden_states: paddle.Tensor,
+        combine_overlap_handle: dict | None = None,
+        async_finish: bool = False,
+    ):
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/moe_layer.py:703]: OK,我改下。
